runtime/lua/vim/lsp/document_highlight.lua

```diff
@@ -0,0 +1,430 @@
+        raw.range['start'].line,
+        util._get_line_byte_from_position(self.bufnr, raw.range['start'], client.offset_encoding),
+        raw.range['end'].line,
+        util._get_line_byte_from_position(self.bufnr, raw.range['end'], client.offset_encoding),
+      },
+    }
+  end
+  table.sort(highlights, function(a, b)
+    return Range.cmp_pos.lt(a.range[3], a.range[4], b.range[1], b.range[2])
+  end)
+
+  state.highlights = highlights
+  state.version = self.version
+
+  api.nvim__redraw({ buf = self.bufnr, valid = true })
+end
+
+---@package
+---@param client_id? integer
+function State:request(client_id)
+  local method = ms.textDocument_documentHighlight
+
+  for id in pairs(client_id and { client_id } or self.client_state) do
+    local client = assert(vim.lsp.get_client_by_id(id))
+    local params = util.make_position_params(0, client.offset_encoding)
+
+    util._cancel_requests({
+      bufnr = self.bufnr,
+      clients = { client },
+      method = method,
+      type = 'pending',
+    })
+
+    client:request(method, params, function(...)
+      self:handler(...)
+    end, self.bufnr)
+  end
+end
+
+---@param kind lsp.DocumentHighlightKind
+---@return string
+local function kind_hl(kind)
+  if kind == protocol.DocumentHighlightKind.Read then
+    return 'LspReferenceRead'
+  elseif kind == protocol.DocumentHighlightKind.Write then
+    return 'LspReferenceWrite'
+  else -- kind == 1 also the default
+    return 'LspReferenceText'
+  end
+end
+
+---@package
+---@param toprow integer
+---@param botrow integer
+function State:on_win(toprow, botrow)
+  for _, state in pairs(self.client_state) do
+    -- Buffer changes may invalidate the original highlight ranges,
+    -- never set outdated highlights as extmarks.
+    if state.version == self.version then
+      api.nvim_buf_clear_namespace(self.bufnr, state.namespace, 0, -1)
+
+      -- Only set extmarks for visible lines
+      local highlights = state.highlights
+      local first = lower_bound(highlights, toprow, 0, 1, #highlights + 1)
+      local last = upper_bound(highlights, botrow, math.huge, first, #highlights + 1) - 1
+
+      for i = first, last do
+        local row, col, end_row, end_col = Range.unpack4(highlights[i].range)
+
+        api.nvim_buf_set_extmark(self.bufnr, state.namespace, row, col, {
+          end_row = end_row,
+          end_col = end_col,
+          hl_group = kind_hl(highlights[i].kind),
+          -- Although we want to avoid
+          -- showing outdated document highlights when the cursor moves,
+          -- updating highlights after a document change
+          -- requires waiting for the server's response.
+          -- This delay can cause flickering, so we don't use ephemeral extmarks here.
+          ephemeral = false,
+        })
+      end
+    end
+  end
+end
+
+local namespace = api.nvim_create_namespace('nvim.lsp.document_highlight')
+api.nvim_set_decoration_provider(namespace, {
+  on_win = function(_, _, bufnr, toprow, botrow)
```

Comment:
generally want to avoid setting up global handlers on module-load. should this be in enable() somewhere ?